Polygon_repair_2/include/CGAL/Polygon_repair_2/Polygon_repair_2.h

```diff
@@ -0,0 +1,346 @@
+      for (int opposite_vertex = 0; opposite_vertex < 3; ++opposite_vertex) {
+        if (face->label() == face->neighbor(opposite_vertex)->label()) continue;
+        nesting[-face->label()-2].insert(face->neighbor(opposite_vertex)->label());
+      }
+    } 
+
+    // int hole_label = -2;
+    // for (auto const &hole: nesting) {
+    //   std::cout << "Hole " << hole_label-- << " contained in polygon(s): ";
+    //   for (auto const &polygon: hole) {
+    //     std::cout << polygon << " ";
+    //   } std::cout << std::endl;
+    // }
+  }
+
+  // Reconstruct multipolygon based on the triangles labelled as inside the polygon
+  void reconstruct_multipolygon() {
+    mp.clear();
+    std::vector<Polygon_2<Kernel, PolygonContainer>> polygons;
+    std::vector<std::set<Polygon_2<Kernel, PolygonContainer>, Polygon_less>> holes; // holes are ordered
+    for (int i = 0; i < number_of_polygons; ++i) {
+      polygons.emplace_back();
+      holes.emplace_back();
+    }
+
+    for (auto const face: t.all_face_handles()) {
+      face->processed() = false;
+    } for (auto const &face: t.finite_face_handles()) {
+      if (face->label() == -1) continue; // exterior triangle
+      if (face->label() < -1 && nesting[-face->label()-2].size() > 1) continue; // exterior triangle
+      if (face->processed()) continue; // already reconstructed
+      for (int opposite_vertex = 0; opposite_vertex < 3; ++opposite_vertex) {
+        if (face->label() != face->neighbor(opposite_vertex)->label()) {
+
+          std::list<typename Kernel::Point_2> ring;
+          reconstruct_ring(ring, face, opposite_vertex);
+          if (face->label() > 0) {
+            polygons[face->label()-1].insert(polygons[face->label()-1].vertices_end(),
+                                             ring.begin(), ring.end());
+          } else {
+            int hole_nesting = *(nesting[-face->label()-2].begin());
+            // std::cout << "Hole: " << face->label() << " -> item " << -face->label()-2 << " -> in polygon " << hole_nesting << std::endl;
+            ring.push_back(ring.front());
+            ring.pop_front();
+            holes[hole_nesting-1].insert(Polygon_2<Kernel, PolygonContainer>(ring.rbegin(), ring.rend()));
+          }
+
+          std::list<typename Triangulation::Face_handle> to_check;
+          to_check.push_back(face);
+          while (!to_check.empty()) {
+            for (int neighbour = 0; neighbour < 3; ++neighbour) {
+              if (to_check.front()->label() == to_check.front()->neighbor(neighbour)->label() &&
+                  !to_check.front()->neighbor(neighbour)->processed()) {
+                to_check.push_back(to_check.front()->neighbor(neighbour));
+              }
+            } to_check.front()->processed() = true;
+            to_check.pop_front();
+          }
+
+          break;
+        }
+      }
+    }
+
+    // Create polygons with holes and put in multipolygon
```

Comment:
I think globally there is a lot of copying polygons going on.  In the optimisation phase of the GSoC  we should check where we can use move semantics.